packages/fxa-settings/src/pages/Signup/ConfirmSignupCode/container.tsx

```diff
@@ -0,0 +1,115 @@
+import { RouteComponentProps, useLocation } from '@reach/router';
+import { currentAccount } from '../../../lib/cache';
+import { useFinishOAuthFlowHandler } from '../../../lib/oauth/hooks';
+import {
+  Integration,
+  isOAuthIntegration,
+  useAuthClient,
+} from '../../../models';
+import AppLayout from '../../../components/AppLayout';
+import CardHeader from '../../../components/CardHeader';
+import ConfirmSignupCode from '.';
+import { hardNavigateToContentServer } from 'fxa-react/lib/utils';
+import LoadingSpinner from 'fxa-react/components/LoadingSpinner';
+import { LocationState } from './interfaces';
+
+const SignupConfirmCodeContainer = ({
+  integration,
+}: {
+  integration: Integration;
+} & RouteComponentProps) => {
+  const authClient = useAuthClient();
+  const location = useLocation() as ReturnType<typeof useLocation> & {
+    state: LocationState;
+  };
+  const {
+    selectedNewsletterSlugs: newsletterSlugs,
+    keyFetchToken,
+    unwrapBKey,
+  } = location.state || {};
+
+  const { finishOAuthFlowHandler, oAuthDataError } = useFinishOAuthFlowHandler(
+    authClient,
+    integration
+  );
+
+  // TODO: UX for this, FXA-8106
+  if (oAuthDataError) {
+    return (
+      <AppLayout>
+        <CardHeader
+          headingText="Unexpected error"
+          headingTextFtlId="auth-error-999"
+        />
+      </AppLayout>
+    );
+  }
+
+  /* Users who reach this page should have account data set in localStorage.
+   * Account data is persisted local storage after creating an (unverified) account
+   * and after sign in. Users may also have localStorage set by the browser if
+   * they are logged in.
+
+   * The session token from local storage is required to verify the signup code.
+   * If this page is reached without an account in localStorage (e.g., directly from
+   * URL), redirect to `/`.
+
+   * TOOD: when we pull the account.verifySession call into the container component,
+   * ensure we're only reading from localStorage once. `sessionToken()` also reads from
+   * localStorage. */
+  const storedLocalAccount = currentAccount();
+  const { email, sessionToken, uid } = storedLocalAccount;
+
+  // Users in this state should never reach this as they should see the Backbone version
+  // of this page until until we convert signin to React, but guard against anyway. See
+  // comment in `router.js` for this route for more info.
+  if (isOAuthIntegration(integration) && (!keyFetchToken || !unwrapBKey)) {
+    const params = new URLSearchParams(location.search);
+    // Remove `emailFromContent` since we pass that when coming
+    // from content-server to Backbone, see Signup container component
+    // for more info.
+    params.delete('emailFromContent');
+    hardNavigateToContentServer(`/?${params.toString()}`);
+    hardNavigateToContentServer(`/signin?${location.search}`);
+    return (
+      <LoadingSpinner className="bg-grey-20 flex items-center flex-col justify-center h-screen select-none" />
```

Comment:
This seems nice, but is the spinner ever seen? I guess I will find out when I do some manual testing, but I've  assumed the hard navigate is almost immediately effective.

Reply:
> I've assumed the hard navigate is almost immediately effective

Without a `return`, the rest of the code tries to run. When not returning something, I've ran into something weird that I can't quite remember which may have just been a memory leak, that `isHardNavigate` in another file had fixed.